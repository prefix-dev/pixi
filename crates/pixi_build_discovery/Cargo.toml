[package]
authors.workspace = true
description = "Discover the build backend from a source tree"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_build_discovery"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
dunce = { workspace = true }
itertools = { workspace = true }
miette = { workspace = true }
pathdiff = { workspace = true }
serde = { workspace = true, optional = true, features = ["derive"] }
serde_json = { workspace = true }
thiserror = { workspace = true }
url = { workspace = true }

rattler_conda_types = { workspace = true }

pixi_build_type_conversions = { workspace = true }
pixi_build_types = { workspace = true }
pixi_manifest = { workspace = true }
pixi_spec = { workspace = true }
pixi_spec_containers = { workspace = true }

[dev-dependencies]
insta = { workspace = true, features = [
  "yaml",
  "filters",
  "glob",
  "redactions",
] }
pixi_build_discovery = { path = ".", features = ["serde"] }
pixi_test_utils = { workspace = true }
