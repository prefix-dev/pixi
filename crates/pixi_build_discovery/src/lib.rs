//! A crate that is responsible for discovering the build backend that should
//! be used for a particular source tree.
//!
//! # Example
//!
//! ```rust,no_run
//! use pixi_build_discovery::{DiscoveredBackend, EnabledProtocols};
//! use rattler_conda_types::ChannelConfig;
//! use std::path::{Path, PathBuf};
//!
//! // Set up the channel configuration (typically from your environment)
//! let channel_config = ChannelConfig::default_with_root_dir(PathBuf::from("."));
//!
//! // Enable all protocols (default)
//! let enabled_protocols = EnabledProtocols::default();
//!
//! // Path to the source directory or manifest
//! let source_path = Path::new("/path/to/project");
//!
//! // Attempt to discover the backend
//! match DiscoveredBackend::discover(source_path, &channel_config, &enabled_protocols) {
//!     Ok(backend) => println!("Discovered backend: {:?}", backend.backend_spec),
//!     Err(e) => eprintln!("Failed to discover backend: {e}"),
//! }
//! ```
//!
//! This crate provides types and logic to identify and describe build backends for projects
//! using pixi, or rattler-build based workflows. It supports discovery from both `pixi.toml`
//! and `recipe.yaml` manifests, and can be extended to support additional protocols.

mod backend_spec;
mod discovery;

pub use backend_spec::{
    BackendSpec, CommandSpec, EnvironmentSpec, JsonRpcBackendSpec, SystemCommandSpec,
};
pub use discovery::{
    BackendInitializationParams, DiscoveredBackend, DiscoveryError, EnabledProtocols,
};
