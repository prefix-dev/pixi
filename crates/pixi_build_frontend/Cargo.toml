[package]
authors.workspace = true
description = "Sets up environments to use for building of pixi.toml files"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_build_frontend"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
dashmap = { workspace = true }
fs-err = { workspace = true }
futures = { workspace = true }
itertools = { workspace = true }
jsonrpsee = { workspace = true, features = ["client"] }
miette = { workspace = true, features = ["fancy", "serde"] }
pixi_config = { workspace = true }
pixi_consts = { workspace = true }
pixi_manifest = { workspace = true }
pixi_progress = { workspace = true }
pixi_spec = { workspace = true }
pixi_utils = { workspace = true, features = ["rustls-tls"] }
rattler = { workspace = true }
rattler_conda_types = { workspace = true }
rattler_repodata_gateway = { workspace = true, features = ["gateway"] }
rattler_shell = { workspace = true }
rattler_solve = { workspace = true, features = ["resolvo"] }
rattler_virtual_packages = { workspace = true }
regex = { workspace = true }
reqwest-middleware = { workspace = true }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
serde_with = { workspace = true }
serde_yaml = { workspace = true }
sha1 = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["process", "io-std"] }
tokio-util = { workspace = true, features = ["codec"] }
tracing = { workspace = true }
url = "2.5.0"
which = { workspace = true }

pixi_build_types = { path = "../pixi_build_types" }


[dev-dependencies]
bytes = { workspace = true }
insta = { workspace = true, features = ["yaml", "filters"] }
rstest = { workspace = true }
tempfile = { workspace = true }
tokio = { workspace = true, features = [
  "process",
  "io-std",
  "rt-multi-thread",
] }
tokio-stream = { workspace = true }
tokio-util = { workspace = true, features = ["io"] }
