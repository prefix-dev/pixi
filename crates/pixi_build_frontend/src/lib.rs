mod build_frontend;
mod jsonrpc;
pub mod protocol;
mod protocols;

use std::fmt::{Debug, Formatter};

pub(crate) use protocols::builders::{conda_protocol, pixi_protocol, rattler_build_protocol};

mod protocol_builder;
mod reporters;
mod tool;

use std::path::PathBuf;

pub use build_frontend::{BuildFrontend, BuildFrontendError};
use rattler_conda_types::MatchSpec;
pub use reporters::{CondaBuildReporter, CondaMetadataReporter};
pub use reporters::{NoopCondaBuildReporter, NoopCondaMetadataReporter};
use tokio::io::{AsyncRead, AsyncWrite};
pub use tool::{IsolatedToolSpec, SystemToolSpec, ToolContext, ToolSpec};
use url::Url;

pub use crate::protocol::Protocol;

pub use protocol_builder::EnabledProtocols;

#[derive(Debug)]
pub enum BackendOverride {
    /// Overrwide the backend with a specific tool.
    Spec(MatchSpec),

    /// Overwrite the backend with a specific tool.
    System(String),

    /// Use the given IO for the backend.
    Io(InProcessBackend),
}

impl From<InProcessBackend> for BackendOverride {
    fn from(value: InProcessBackend) -> Self {
        Self::Io(value)
    }
}

impl From<InProcessBackend> for Option<BackendOverride> {
    fn from(value: InProcessBackend) -> Self {
        Some(value.into())
    }
}

/// A backend communication protocol that can run in the same process.
pub struct InProcessBackend {
    pub rpc_in: Box<dyn AsyncRead + Send + Sync + Unpin>,
    pub rpc_out: Box<dyn AsyncWrite + Send + Sync + Unpin>,
}

impl Debug for InProcessBackend {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("InProcessBackend").finish()
    }
}

#[derive(Debug)]
pub struct SetupRequest {
    /// The source directory that contains the source package.
    pub source_dir: PathBuf,

    /// Overrides for the build tool.
    pub build_tool_override: Option<BackendOverride>,

    /// Identifier for the rest of the requests
    /// This is used to identify the requests that belong to the same build.
    pub build_id: usize,
}

#[derive(Debug)]
pub struct BuildOutput {
    /// Paths to the built artifacts.
    pub artifacts: Vec<PathBuf>,
}

#[derive(Debug)]
pub struct CondaMetadataRequest {
    /// The base urls of the channels to use.
    pub channel_base_urls: Vec<Url>,
}
