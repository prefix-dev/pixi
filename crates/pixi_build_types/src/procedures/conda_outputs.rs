//! Describes the `conda/outputs` request and its parameters.
//!
//! This request is used to compute all the outputs that a particular backend
//! can provide. It returns the identifiable metadata of the outputs, including
//! the dependencies required to be able to build them.
//!
//! This API was introduced in Pixi Build API version 1.

use std::{
    collections::{BTreeMap, BTreeSet},
    path::PathBuf,
};

use ordermap::OrderSet;
use rattler_conda_types::{ChannelUrl, NoArchType, PackageName, Platform, VersionWithSource};
use serde::{Deserialize, Serialize};
use serde_with::serde_as;

use crate::{BinaryPackageSpecV1, PackageSpecV1, project_model::NamedSpecV1};

pub const METHOD_NAME: &str = "conda/outputs";

/// Parameters for the `conda/outputs` request.
///
/// The result of this request should be a list of packages that can be built by
/// this particular source dependency.
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputsParams {
    /// The canonical channel URLs that define where dependencies will be
    /// fetched from. Although this information is not immediately useful for
    /// the backend, the backend may choose to generate a different recipe based
    /// on the channels.
    #[serde(default)]
    pub channels: Vec<ChannelUrl>,

    /// The native platform for which the outputs should be computed.
    ///
    /// This is usually the same platform as the platform on which the backend
    /// is running but when cross-compiling this could be different.
    pub host_platform: Platform,

    /// The platform for which any build tools will be installed.
    ///
    /// This is usually the same platform as the platform on which the backend
    /// is running but when cross-compiling this could be different.
    pub build_platform: Platform,

    /// The possible variants by the pixi workspace.
    pub variant_configuration: Option<BTreeMap<String, Vec<String>>>,

    /// A directory that can be used by the backend to store files for
    /// subsequent requests. This directory is unique for each separate source
    /// dependency.
    ///
    /// The directory may not yet exist.
    pub work_directory: PathBuf,
}

/// Contains the result of the `conda/outputs` request.
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputsResult {
    /// Metadata of all the packages that can be built.
    pub outputs: Vec<CondaOutput>,

    /// The files that were read as part of the computation. These files are
    /// hashed and stored in the lock-file. If the files change, the
    /// lock-file will be invalidated.
    pub input_globs: BTreeSet<String>,
}

#[serde_as]
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutput {
    /// The identifier of the output.
    pub metadata: CondaOutputMetadata,

    /// The build dependencies of the package. These refer to the packages that
    /// should be installed in the "build" environment. The build environment
    /// contains packages for the current architecture that can be used to run
    /// tools on the current machine like compilers, code generators, etc.
    pub build_dependencies: Option<CondaOutputDependencies>,

    /// The "host" dependencies of the package. These refer to the package that
    /// should be installed to be able to refer to them from the build process
    /// but not run them. They are installed for the "target" architecture (see
    /// subdir) or for the current architecture if the target is `noarch`.
    ///
    /// For C++ packages these would be libraries to link against.
    pub host_dependencies: Option<CondaOutputDependencies>,

    /// The dependencies for the run environment of the package. These
    /// dependencies are installed at runtime when this particular package is
    /// also installed.
    pub run_dependencies: CondaOutputDependencies,

    /// Describes which run-exports should be ignored for this package.
    pub ignore_run_exports: CondaOutputIgnoreRunExports,

    /// The run exports of this particular output.
    pub run_exports: CondaOutputRunExports,

    /// A cache that might be shared between multiple outputs based on the
    /// contents of the cache.
    /// TODO: No yet supported, see commented `CondaCacheMetadata` below.
    // pub cache: Option<CondaCacheMetadata>,

    /// Explicit input globs for this specific output. If this is `None`,
    /// [`CondaOutputsResult::input_globs`] will be used.
    pub input_globs: Option<BTreeSet<String>>,
}

#[serde_as]
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputMetadata {
    /// The name of the package.
    pub name: PackageName,

    /// The version of the package.
    pub version: VersionWithSource,

    /// The build hash of the package.
    pub build: String,

    /// The build number of the package.
    pub build_number: u64,

    /// The subdir or platform
    pub subdir: Platform,

    /// The license of the package
    pub license: Option<String>,

    /// The license family of the package
    pub license_family: Option<String>,

    /// The noarch type of the package
    pub noarch: NoArchType,

    /// Any PURL (Package URL) that is associated with this package.
    pub purls: Option<OrderSet<rattler_conda_types::PackageUrl>>,

    /// Optionally, a path within the environment of the site-packages
    /// directory. This field is only present for "python" interpreter
    /// packages. This field was introduced with <https://github.com/conda/ceps/blob/main/cep-17.md>.
    pub python_site_packages_path: Option<String>,

    /// The variants that were used for this output.
    pub variant: BTreeMap<String, String>,
}

/// Describes dependencies, constraints and source dependencies for a particular
/// environment.
#[serde_as]
#[derive(Debug, Default, Serialize, Deserialize, Clone, Hash, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputDependencies {
    /// A list of matchspecs that describe the dependencies of a particular
    /// environment.
    pub depends: Vec<NamedSpecV1<PackageSpecV1>>,

    /// Additional constraints that apply to the environment in which the
    /// dependencies are solved. Constraints are represented as matchspecs.
    pub constraints: Vec<NamedSpecV1<BinaryPackageSpecV1>>,
}

/// Describes which run-exports should be ignored for a particular output.
#[serde_as]
#[derive(Debug, Default, Serialize, Deserialize, Clone, Hash, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputIgnoreRunExports {
    /// Run exports to ignore by name of the package that is exported
    pub by_name: Vec<PackageName>,

    /// Run exports to ignore by the package that applies them
    pub from_package: Vec<PackageName>,
}

#[serde_as]
#[derive(Debug, Default, Deserialize, Serialize, Clone, Hash, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CondaOutputRunExports {
    /// weak run exports apply a dependency from host to run
    pub weak: Vec<NamedSpecV1<PackageSpecV1>>,

    /// strong run exports apply a dependency from build to host and run
    pub strong: Vec<NamedSpecV1<PackageSpecV1>>,

    /// noarch run exports apply a run export only to noarch packages (other run
    /// exports are ignored) for example, python uses this to apply a
    /// dependency on python to all noarch packages, but not to
    /// the python_abi package
    pub noarch: Vec<NamedSpecV1<PackageSpecV1>>,

    /// weak constrains apply a constrain dependency from host to run
    pub weak_constrains: Vec<NamedSpecV1<BinaryPackageSpecV1>>,

    /// strong constrains apply a constrain dependency from build to host and
    /// run
    pub strong_constrains: Vec<NamedSpecV1<BinaryPackageSpecV1>>,
}

// TODO: Multi-output caching is not yet supported.
// #[serde_as]
// #[derive(Debug, Serialize, Deserialize, Clone, Hash, Eq, PartialEq)]
// #[serde(rename_all = "camelCase")]
// pub struct CondaCacheMetadata {
//     /// An optional name
//     pub name: Option<String>,
//
//     /// The build dependencies of the package. These refer to the packages that
//     /// should be installed in the "build" environment. The build environment
//     /// contains packages for the current architecture that can be used to run
//     /// tools on the current machine like compilers, code generators, etc.
//     pub build_dependencies: Option<CondaOutputDependencies>,
//
//     /// The "host" dependencies of the package. These refer to the package that
//     /// should be installed to be able to refer to them from the build process
//     /// but not run them. They are installed for the "target" architecture (see
//     /// subdir) or for the current architecture if the target is `noarch`.
//     /// For C++ packages these would be libraries to link against.
//     pub host_dependencies: Option<CondaOutputDependencies>,
//
//     /// Describes which run-exports should be ignored for this package.
//     pub ignore_run_exports: CondaOutputIgnoreRunExports,
// }
