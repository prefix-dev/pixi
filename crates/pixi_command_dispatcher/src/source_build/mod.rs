use std::{
    collections::{BTreeMap, BTreeSet},
    path::{Path, PathBuf},
};

use miette::Diagnostic;
use pixi_build_discovery::{DiscoveredBackend, EnabledProtocols};
use pixi_build_frontend::Backend;
use pixi_build_types::procedures::conda_outputs::CondaOutputsParams;
use pixi_record::{PinnedSourceSpec, PixiRecord};
use pixi_spec::{SourceAnchor, SourceSpec};
use rattler_conda_types::{
    ChannelConfig, ChannelUrl, InvalidPackageNameError, Platform, prefix::Prefix,
};
use thiserror::Error;
use tracing::instrument;

use crate::{
    BackendBuiltSource, BackendSourceBuildError, BackendSourceBuildMethod,
    BackendSourceBuildPrefix, BackendSourceBuildSpec, BackendSourceBuildV0Method,
    BackendSourceBuildV1Method, BuildEnvironment, CommandDispatcher, CommandDispatcherError,
    CommandDispatcherErrorResultExt, InstallPixiEnvironmentError, InstallPixiEnvironmentSpec,
    InstantiateBackendError, InstantiateBackendSpec, PixiEnvironmentSpec,
    SolvePixiEnvironmentError, SourceCheckout, SourceCheckoutError,
    build::{
        Dependencies, DependenciesError, MoveError, SourceRecordOrCheckout, WorkDirKey, move_file,
    },
    package_identifier::PackageIdentifier,
};

/// Describes all parameters required to build a conda package from a pixi
/// source package.
///
/// This task prepares the build environment for a source build and then
/// delegates the actual build to the backend through the
/// [`BackendSourceBuildSpec`]. This allows preparation (installing host, build,
/// envs) to progress concurrently while the actual building of the package can
/// be done serially.
#[derive(Debug, Clone, serde::Serialize)]
pub struct SourceBuildSpec {
    /// The source to build
    pub package: PackageIdentifier,

    /// The location of the source code to build.
    pub source: PinnedSourceSpec,

    /// The channel configuration to use when resolving metadata
    pub channel_config: ChannelConfig,

    /// The channels to use for solving.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub channels: Vec<ChannelUrl>,

    /// Information about host platform on which the package is build. Note that
    /// a package might be targeting noarch in which case the host platform
    /// should be used.
    ///
    /// If this field is omitted the build backend will use the current
    /// platform.
    pub build_environment: BuildEnvironment,

    /// Variant configuration
    pub variants: Option<BTreeMap<String, Vec<String>>>,

    /// The directory where to place the built package.
    pub output_directory: Option<PathBuf>,

    /// The working directory to use for the build. If this is `None` a
    /// deterministic workspace local directory will be used.
    pub work_directory: Option<PathBuf>,

    /// Whether the build directory should be cleared before building.
    pub clean: bool,

    /// The protocols that are enabled for this source
    #[serde(skip_serializing_if = "crate::is_default")]
    pub enabled_protocols: EnabledProtocols,
}

pub struct BuiltSource {
    /// The source checkout that was built
    pub source: SourceCheckout,

    /// The location on disk where the built package is located.
    pub output_file: PathBuf,

    /// The globs that were used as input to the build. Use these for
    /// re-verifying the build.
    pub input_globs: BTreeSet<String>,
}

impl SourceBuildSpec {
    #[instrument(skip_all, fields(
        source = %self.source,
        subdir = %self.package.subdir,
        name = %self.package.name.as_normalized(),
        version = %self.package.version,
        build = %self.package.build))]
    pub(crate) async fn build(
        mut self,
        command_dispatcher: CommandDispatcher,
    ) -> Result<BuiltSource, CommandDispatcherError<SourceBuildError>> {
        tracing::debug!("Building package for source spec: {}", self.source);

        // Check out the source code.
        let source_checkout = command_dispatcher
            .checkout_pinned_source(self.source.clone())
            .await
            .map_err_with(SourceBuildError::SourceCheckout)?;

        // Discover information about the build backend from the source code.
        let discovered_backend = DiscoveredBackend::discover(
            &source_checkout.path,
            &self.channel_config,
            &self.enabled_protocols,
        )
        .map_err(SourceBuildError::Discovery)
        .map_err(CommandDispatcherError::Failed)?;

        // Instantiate the backend with the discovered information.
        let backend = command_dispatcher
            .instantiate_backend(InstantiateBackendSpec {
                backend_spec: discovered_backend.backend_spec,
                init_params: discovered_backend.init_params,
                channel_config: self.channel_config.clone(),
                enabled_protocols: self.enabled_protocols.clone(),
            })
            .await
            .map_err_with(SourceBuildError::Initialize)?;

        // Determine the working directory for the build.
        let work_directory = match std::mem::take(&mut self.work_directory) {
            Some(work_directory) => work_directory,
            None => command_dispatcher.cache_dirs().working_dirs().join(
                WorkDirKey {
                    source: SourceRecordOrCheckout::Record {
                        pinned: self.source.clone(),
                        package_name: self.package.name.clone(),
                    },
                    host_platform: self.build_environment.host_platform,
                    build_backend: backend.identifier().to_string(),
                }
                .key(),
            ),
        };

        // Clean the working directory if requested.
        if self.clean {
            if let Err(err) = fs_err::remove_dir_all(&work_directory) {
                return Err(CommandDispatcherError::Failed(
                    SourceBuildError::CleanWorkingDirectory(work_directory, err),
                ));
            }
        }

        // Build the package based on the support backend capabilities.
        let output_directory = self.output_directory.clone();
        let mut built_source = if backend.capabilities().provides_conda_build_v1() {
            let built_package = self
                .build_v1(command_dispatcher, backend, work_directory)
                .await?;

            BuiltSource {
                source: source_checkout,
                input_globs: built_package.input_globs,
                output_file: built_package.output_file,
            }
        } else {
            let built_package = self
                .build_v0(command_dispatcher, backend, work_directory)
                .await?;

            BuiltSource {
                source: source_checkout,
                input_globs: built_package.input_globs,
                output_file: built_package.output_file,
            }
        };

        // Make sure the package resides in the output directory that was requested.
        if let Some(output_directory) = output_directory {
            // Create the output directory if it does not exist.
            fs_err::create_dir_all(&output_directory).map_err(|err| {
                CommandDispatcherError::Failed(SourceBuildError::CreateOutputDirectory(err))
            })?;

            // At this point, the directory should exist, so we can canonicalize the path.
            let output_directory = fs_err::canonicalize(&output_directory)
                .map_err(CommandDispatcherError::Failed)
                .map_err_with(SourceBuildError::CreateOutputDirectory)?;

            // The output file should also exist.
            let output_file = match fs_err::canonicalize(&built_source.output_file) {
                Ok(output_file) => output_file,
                Err(_err) => {
                    return Err(CommandDispatcherError::Failed(
                        SourceBuildError::MissingOutputFile(built_source.output_file),
                    ));
                }
            };

            if output_file.parent() != Some(&output_directory) {
                // Take the file name of the file and move it to the output directory.
                let file_name = built_source
                    .output_file
                    .file_name()
                    .expect("the build backend did not return a file name");
                let destination = output_directory.join(file_name);
                if let Err(err) = move_file(&output_file, &destination) {
                    return Err(CommandDispatcherError::Failed(SourceBuildError::Move(
                        output_file,
                        output_directory,
                        err,
                    )));
                }
                built_source.output_file = destination;
            }
        }

        Ok(built_source)
    }

    async fn build_v0(
        self,
        command_dispatcher: CommandDispatcher,
        backend: Backend,
        work_directory: PathBuf,
    ) -> Result<BackendBuiltSource, CommandDispatcherError<SourceBuildError>> {
        command_dispatcher
            .backend_source_build(BackendSourceBuildSpec {
                backend,
                package: self.package,
                source: self.source,
                work_directory,
                method: BackendSourceBuildMethod::BuildV0(BackendSourceBuildV0Method {
                    channel_config: self.channel_config,
                    channels: self.channels,
                    build_environment: self.build_environment,
                    variants: self.variants,
                    output_directory: self.output_directory,
                }),
            })
            .await
            .map_err_with(SourceBuildError::from)
    }

    async fn build_v1(
        self,
        command_dispatcher: CommandDispatcher,
        backend: Backend,
        work_directory: PathBuf,
    ) -> Result<BackendBuiltSource, CommandDispatcherError<SourceBuildError>> {
        let source_anchor = SourceAnchor::from(SourceSpec::from(self.source.clone()));
        let host_platform = self.build_environment.host_platform;
        let build_platform = self.build_environment.build_platform;

        // Request the metadata from the backend.
        // TODO: Can we somehow cache this metadata?
        let outputs = backend
            .conda_outputs(CondaOutputsParams {
                host_platform,
                build_platform,
                variant_configuration: self.variants.clone(),
                work_directory: work_directory.clone(),
            })
            .await
            .map_err(BackendSourceBuildError::BuildError)
            .map_err(SourceBuildError::from)
            .map_err(CommandDispatcherError::Failed)?;

        // Find the output that we want to build.
        let output = outputs
            .outputs
            .into_iter()
            .find(|output| {
                output.metadata.name == self.package.name
                    && output.metadata.version == self.package.version
                    && output.metadata.build == self.package.build
                    && output.metadata.subdir.as_str() == self.package.subdir
            })
            .ok_or_else(|| {
                CommandDispatcherError::Failed(SourceBuildError::MissingOutput {
                    subdir: self.package.subdir.clone(),
                    name: self.package.name.as_normalized().to_string(),
                    version: self.package.version.to_string(),
                    build: self.package.build.clone(),
                })
            })?;

        // Determine final directories for everything.
        let directories = Directories::new(&work_directory, host_platform);

        // Solve the build environment.
        let build_dependencies = output
            .build_dependencies
            .as_ref()
            .map(|deps| Dependencies::new(deps, Some(source_anchor.clone())))
            .transpose()
            .map_err(SourceBuildError::from)
            .map_err(CommandDispatcherError::Failed)?
            .unwrap_or_default();
        let build_records = self
            .solve_dependencies(
                format!("{} (build)", self.package.name.as_source()),
                &command_dispatcher,
                build_dependencies.clone(),
                self.build_environment.to_build_from_build(),
            )
            .await
            .map_err_with(Box::new)
            .map_err_with(SourceBuildError::SolveBuildEnvironment)?;
        let build_run_exports =
            build_dependencies.extract_run_exports(&build_records, &output.ignore_run_exports);

        // Solve the host environment for the output.
        let host_dependencies = output
            .host_dependencies
            .as_ref()
            .map(|deps| Dependencies::new(deps, Some(source_anchor.clone())))
            .transpose()
            .map_err(SourceBuildError::from)
            .map_err(CommandDispatcherError::Failed)?
            .unwrap_or_default()
            // Extend with the run exports from the build environment.
            .extend_with_run_exports_from_build(&build_run_exports);
        let host_records = self
            .solve_dependencies(
                format!("{} (host)", self.package.name.as_source()),
                &command_dispatcher,
                host_dependencies.clone(),
                self.build_environment.clone(),
            )
            .await
            .map_err_with(Box::new)
            .map_err_with(SourceBuildError::SolveBuildEnvironment)?;

        // Install the build environment
        let _build_prefix = command_dispatcher
            .install_pixi_environment(InstallPixiEnvironmentSpec {
                name: format!("{} (build)", self.package.name.as_source()),
                records: build_records,
                prefix: Prefix::create(&directories.build_prefix)
                    .map_err(SourceBuildError::CreateBuildEnvironmentDirectory)
                    .map_err(CommandDispatcherError::Failed)?,
                installed: None,
                build_environment: self.build_environment.to_build_from_build(),
                force_reinstall: Default::default(),
                channels: self.channels.clone(),
                channel_config: self.channel_config.clone(),
                variants: self.variants.clone(),
                enabled_protocols: self.enabled_protocols.clone(),
            })
            .await
            .map_err_with(Box::new)
            .map_err_with(SourceBuildError::InstallBuildEnvironment)?;

        // Install the host environment.
        let _host_prefix = command_dispatcher
            .install_pixi_environment(InstallPixiEnvironmentSpec {
                name: format!("{} (host)", self.package.name.as_source()),
                records: host_records,
                prefix: Prefix::create(&directories.host_prefix)
                    .map_err(SourceBuildError::CreateBuildEnvironmentDirectory)
                    .map_err(CommandDispatcherError::Failed)?,
                installed: None,
                build_environment: self.build_environment.to_build_from_build(),
                force_reinstall: Default::default(),
                channels: self.channels.clone(),
                channel_config: self.channel_config.clone(),
                variants: self.variants.clone(),
                enabled_protocols: self.enabled_protocols.clone(),
            })
            .await
            .map_err_with(Box::new)
            .map_err_with(SourceBuildError::InstallBuildEnvironment)?;

        command_dispatcher
            .backend_source_build(BackendSourceBuildSpec {
                backend,
                package: self.package,
                source: self.source,
                work_directory,
                method: BackendSourceBuildMethod::BuildV1(BackendSourceBuildV1Method {
                    build_prefix: BackendSourceBuildPrefix {
                        platform: self.build_environment.build_platform,
                        prefix: directories.build_prefix,
                    },
                    host_prefix: BackendSourceBuildPrefix {
                        platform: self.build_environment.host_platform,
                        prefix: directories.host_prefix,
                    },
                    variant: output.metadata.variant,
                    output_directory: self.output_directory,
                }),
            })
            .await
            .map_err_with(SourceBuildError::from)
    }

    async fn solve_dependencies(
        &self,
        name: String,
        command_dispatcher: &CommandDispatcher,
        dependencies: Dependencies,
        build_environment: BuildEnvironment,
    ) -> Result<Vec<PixiRecord>, CommandDispatcherError<SolvePixiEnvironmentError>> {
        if dependencies.dependencies.is_empty() {
            return Ok(vec![]);
        }
        command_dispatcher
            .solve_pixi_environment(PixiEnvironmentSpec {
                name: Some(name),
                dependencies: dependencies.dependencies,
                constraints: dependencies.constraints,
                installed: vec![], // TODO: To lock build environments, fill this.
                build_environment,
                channels: self.channels.clone(),
                strategy: Default::default(),
                channel_priority: Default::default(),
                exclude_newer: None,
                channel_config: self.channel_config.clone(),
                variants: self.variants.clone(),
                enabled_protocols: self.enabled_protocols.clone(),
            })
            .await
    }
}

pub struct Directories {
    host_prefix: PathBuf,
    build_prefix: PathBuf,
}

impl Directories {
    pub fn new(work_directory: &Path, host_platform: Platform) -> Self {
        const BUILD_DIR: &str = "bld";
        const HOST_ENV_DIR: &str = "host";
        const PLACEHOLDER_TEMPLATE_STR: &str = "_placehold";

        let build_prefix = work_directory.join(BUILD_DIR);
        let host_prefix = if host_platform.is_windows() {
            work_directory.join(HOST_ENV_DIR)
        } else {
            // On non-Windows platforms, the name of the host environment has to be exactly
            // 255 characters long for prefix replacement in rattler build to work
            // correctly. This code constructs a directory name padded with a
            // template string so its exactly 255 characters long.
            //
            // TODO: This is really an implementation detail of how backends are generally
            // implemented, but this code should not really live in pixi.
            const PLACEHOLDER_LENGTH: usize = 255;
            let mut placeholder = String::new();
            while placeholder.len() < PLACEHOLDER_LENGTH {
                placeholder.push_str(PLACEHOLDER_TEMPLATE_STR);
            }
            let placeholder = placeholder
                [0..PLACEHOLDER_LENGTH - work_directory.join(HOST_ENV_DIR).as_os_str().len()]
                .to_string();

            work_directory.join(format!("{HOST_ENV_DIR}{}", placeholder))
        };
        Self {
            host_prefix,
            build_prefix,
        }
    }
}

#[derive(Debug, Error, Diagnostic)]
pub enum SourceBuildError {
    #[error(transparent)]
    #[diagnostic(transparent)]
    SourceCheckout(#[from] SourceCheckoutError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Discovery(#[from] pixi_build_discovery::DiscoveryError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Initialize(#[from] InstantiateBackendError),

    #[error("failed to solve the build environment")]
    SolveBuildEnvironment(
        #[diagnostic_source]
        #[source]
        Box<SolvePixiEnvironmentError>,
    ),

    #[error("failed to solve the host environment")]
    SolveHostEnvironment(
        #[diagnostic_source]
        #[source]
        Box<SolvePixiEnvironmentError>,
    ),

    #[error("failed to create the build environment directory")]
    CreateBuildEnvironmentDirectory(#[source] std::io::Error),

    #[error("failed to create the host environment directory")]
    CreateHostEnvironmentDirectory(#[source] std::io::Error),

    #[error("failed to install the build environment")]
    InstallBuildEnvironment(#[source] Box<InstallPixiEnvironmentError>),

    #[error("failed to install the host environment")]
    InstallHostEnvironment(#[source] Box<InstallPixiEnvironmentError>),

    #[error(
        "The build backend does not provide the requested output: {subdir}/{name}={version}={build}."
    )]
    MissingOutput {
        subdir: String,
        name: String,
        version: String,
        build: String,
    },

    #[error(
        "The build backend returned a path for the build package ({0}), but the path does not exist."
    )]
    MissingOutputFile(PathBuf),

    #[error("backend returned a dependency on an invalid package name: {0}")]
    InvalidPackageName(String, #[source] InvalidPackageNameError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    BackendBuildError(#[from] BackendSourceBuildError),

    #[error("failed to clean the working directory: {0}")]
    CleanWorkingDirectory(PathBuf, #[source] std::io::Error),

    #[error("moving the built package from {0} to the output directory {1} failed")]
    Move(PathBuf, PathBuf, #[source] MoveError),

    #[error("failed to create the output directory")]
    CreateOutputDirectory(#[source] std::io::Error),
}

impl From<DependenciesError> for SourceBuildError {
    fn from(value: DependenciesError) -> Self {
        match value {
            DependenciesError::InvalidPackageName(name, error) => {
                SourceBuildError::InvalidPackageName(name, error)
            }
        }
    }
}
