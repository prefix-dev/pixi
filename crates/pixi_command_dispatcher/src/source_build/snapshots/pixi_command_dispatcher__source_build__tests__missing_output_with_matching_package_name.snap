---
source: crates/pixi_command_dispatcher/src/source_build/mod.rs
expression: error.to_string()
---
The build backend did not provide the requested output for package 'scipy' with variant {"blas":"mkl","python":"3.11"}

Available variants for scipy:
  - {"blas":"mkl","python":"3.10"}  ← 1/2 keys match
  - {"blas":"openblas","python":"3.11"}  ← 1/2 keys match
  - {"blas":"mkl","python":"3.12"}  ← 1/2 keys match
